### PR TITLE
Fix potential ThreadLocal transaction leak [tp-tests]

### PR DESCRIPTION
**Pull request by li-boxuan**
This is a follow-up of #1653, which fixes some ThreadLocal transaction leak
but relies on users explicitly calling graph.tx().close(). This commit fixes
this flaw and now graph.tx().commit() or graph.tx().rollback() should close
all ThreadLocal objects used in JanusGraphBlueprintsGraph and AbstractThreadLocalTransaction.

However, note that we still have ThreadLocal leakage in case the user closes
the graph without committing/rolling back/closing ThreadLocal transactions.
Unfortunately, I don't have an appropriate solution for this at the moment.

Partial fix for #1278

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [ ] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### CODE REVIEW

**Comment by li-boxuan** [comment]
Using the sample test https://gist.github.com/mad/4230d56d73587c11dcf48eb9463ce751 provided by @mad in https://github.com/JanusGraph/janusgraph/pull/1653#issuecomment-508139294, but with a small change (remove L54 "graph.tx().close()"), we can see ThreadLocal leaks using the master version:

image


After this PR, no tx leakage:
image

However, note that we still have ThreadLocal leakage in case the user closes the graph without committing/rolling back/closing ThreadLocal transactions. Unfortunately, I don't have an appropriate solution for this.

File: janusgraph-core/src/main/java/org/janusgraph/core/JanusGraph.java [hunk 1]
```diff
@@ -57,6 +57,11 @@
         reason = "JanusGraph test graph computer (FulgoraGraphComputer) " +
             "currently does not support graph filters but does not throw proper exception because doing so breaks numerous " +
             "tests in gremlin-test ProcessComputerSuite.")
+@Graph.OptOut(
```
Comment (mad) [inline janusgraph-core/src/main/java/org/janusgraph/core/JanusGraph.java:60]: New behavior may fail users code/tests, who reuse transaction listeners within auto tx, so this is breaking change, may be add some notes to upgrade/release notes?

Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/core/JanusGraph.java:60]: That is a good point. Now I am wondering if we should introduce this breaking change, or leave this memory leak as it is (provided there is no other way to fix the leak caused by transactionListeners).

Alternatively, we can let users know that after a transaction is done, they will have to call `graph.tx().addTransactionListener(listener);` explicitly if they want to reuse the listener in the new transaction.

**Review by analytically — APPROVED** [review]

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/JanusGraphBlueprintsGraph.java [hunk 2]
```diff
@@ -287,6 +289,18 @@ public GraphTransaction() {
             super(JanusGraphBlueprintsGraph.this);
         }
 
+        @Override
+        public void commit() {
+            super.commit();
+            doClose();
```
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/JanusGraphBlueprintsGraph.java:295]: This should be placed in a try - finally block to ensure threadlocals are cleaned even if commit/rollback fails.

**Review by li-boxuan — COMMENTED** [review]
I just discovered that the same fix was applied by my coworker 1.5 years ago to fix OOM and it has been running in our production system since then. Therefore, this PR should be safe to merge. Disclaimer: we never use transactionListeners.

File: janusgraph-core/src/main/java/org/janusgraph/core/JanusGraph.java (hunk 1 above)
Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/core/JanusGraph.java:60]: @mad Do you want to review again?

**Review by porunov — COMMENTED** [review]
I never used `transactionListeners` but the changes makes sense to me.
LGTM.

**Pull request merged**